docs/app/pages/components/sections/utilities/color-service/snippets/app.ts

```diff
@@ -0,0 +1,165 @@
+
+                        return `${ precentage }%, Sales ${ item.index + 1 }`;
+                    }
+                },
+                backgroundColor: tooltipBackgroundColor,
+                cornerRadius: 0,
+                displayColors: false
+            } as any
+        };
+
+        this.donutChartColors = [
+            {
+                backgroundColor: [
+                    colorService.getColor('chart1').toRgb(),
+                    colorService.getColor('chart2').toRgb(),
+                    colorService.getColor('chart3').toRgb(),
+                ],
+                hoverBackgroundColor: [
+                    colorService.getColor('chart1').setAlpha(0.3).toRgba(),
+                    colorService.getColor('chart2').setAlpha(0.3).toRgba(),
+                    colorService.getColor('chart3').setAlpha(0.3).toRgba(),
+                ]
+            }
+        ];
+    }
+    
+    changeSalesColor(sales: string, color: string): void {
+        
+          if (!sales || !color) {
+              return;
+              
+          } else {
+        
+            switch (sales) {
+
+                case 'sales1':
+                    this.sales1 = color;
+                    break;
+
+                case 'sales2':
+                    this.sales2 = color;
+                    break;
+                    
+                case 'sales3':
+                    this.sales3 = color;
+            }
+              let colorString = this.colorService.getColor(color.toLowerCase()).toHex();
+              this.setColor(sales, colorString, false);
+          }
+      }
+      
+      setColor(sales: string, color: string, customized: boolean = false): void {
+        
+          if (!sales || !color) {
+              return;
+            
+          } else if (customized && (color.length != 7)) {
+              return;
+            
+          } else {
+            
+              let index;
+            
+              switch (sales) {
+                
+                  case 'sales1':
+                    index = 0;
+                    if (customized) {
+                        this.sales1 = '';
+                    }
+                    break;
+                    
+                  case 'sales2':
+                    index = 1;
+                    if (customized) {
+                        this.sales2 = '';
+                    }
+                    break;
+                    
+                  case 'sales3':
+                    index = 2;
+                    if (customized) {
+                        this.sales3 = '';
+                    }
+              }
+              
+              let newColor = ThemeColor.parse(color);
+              this.donutChartColors[0].backgroundColor[index] = color;
+              this.donutChartColors[0].hoverBackgroundColor[index] = newColor.setAlpha(0.3).toRgba();
+              this.donutChartColors = this.donutChartColors.slice();
```

Comment:
I appreciate that you have taken the time to update the colors individually. When it comes to efficiency in javascript/typescript, there is another thing to consider: the amount of code that the client has to download. Most browsers support array functions like map, reduce, find and so on, which we can use to reduce the amount of code written. (We use polyfills, which inject these functions for browsers which do not support them natively.)

Here are some example functions that might be of use:
````javascript
    colorNamesUpdated() {
        this.salesColorValues = this.salesColorNames.map(colorName => this.colorService.getColor(colorName).toHex());
        this.colorValuesUpdated();
    }
    
    colorValuesUpdated() {
        const chartColors = this.salesColorValues.slice();
        const chartHoverColors = this.salesColorValues.map(color => ThemeColor.parse(color).setAlpha(0.3).toRgba());
        this.donutChartColors = [{
            backgroundColor: chartColors,
            hoverBackgroundColor: chartHoverColors
        }];
    }
````

For these I used two new arrays, which are 2-way bound to the ux-select and input fields. Since the data is stored in a manner which is easily bound to the UI, it is a lot easier to keep everything in sync. In this case, the valueChange/ngModelChange events are used to trigger colorNamesUpdated/colorValuesUpdated, and parameters are not needed because the value is already updated by the Angular binding.